render.c

```diff
@@ -1,69 +1,93 @@
 #include <assert.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
+#include <pixman.h>
 
 #include "buffer.h"
 #include "output-layout.h"
 #include "render.h"
 
-static bool convert_buffer(struct grim_buffer *buffer) {
-	// Formats are little-endian
-	switch (buffer->format) {
+static pixman_format_code_t get_pixman_format(enum wl_shm_format wl_fmt) {
+	switch (wl_fmt) {
 	case WL_SHM_FORMAT_ARGB8888:
+		return PIXMAN_a8r8g8b8;
 	case WL_SHM_FORMAT_XRGB8888:
-		// Natively supported by Cairo
-		return true;
+		return PIXMAN_x8r8g8b8;
 	case WL_SHM_FORMAT_ABGR8888:
-	case WL_SHM_FORMAT_XBGR8888:;
-		// ABGR -> ARGB
-		uint8_t *data = buffer->data;
-		for (int i = 0; i < buffer->height; ++i) {
-			for (int j = 0; j < buffer->width; ++j) {
-				uint32_t *px = (uint32_t *)(data + i * buffer->stride + j * 4);
-				uint8_t a = (*px >> 24) & 0xFF;
-				uint8_t b = (*px >> 16) & 0xFF;
-				uint8_t g = (*px >> 8) & 0xFF;
-				uint8_t r = *px & 0xFF;
-				*px = (a << 24) | (r << 16) | (g << 8) | b;
-			}
-		}
-		if (buffer->format == WL_SHM_FORMAT_ABGR8888) {
-			buffer->format = WL_SHM_FORMAT_ARGB8888;
-		} else {
-			buffer->format = WL_SHM_FORMAT_XRGB8888;
-		}
-		return true;
+		return PIXMAN_a8b8g8r8;
+	case WL_SHM_FORMAT_XBGR8888:
+		return PIXMAN_x8b8g8r8;
+	case WL_SHM_FORMAT_BGRA8888:
+		return PIXMAN_b8g8r8a8;
+	case WL_SHM_FORMAT_BGRX8888:
+		return PIXMAN_b8g8r8x8;
+	case WL_SHM_FORMAT_RGBA8888:
+		return PIXMAN_r8g8b8a8;
+	case WL_SHM_FORMAT_RGBX8888:
+		return PIXMAN_r8g8b8x8;
+	case WL_SHM_FORMAT_ARGB2101010:
+		return PIXMAN_a2r10g10b10;
+	case WL_SHM_FORMAT_ABGR2101010:
+		return PIXMAN_a2b10g10r10;
+	case WL_SHM_FORMAT_XRGB2101010:
+		return PIXMAN_x2r10g10b10;
+	case WL_SHM_FORMAT_XBGR2101010:
+		return PIXMAN_x2b10g10r10;
 	default:
-		fprintf(stderr, "unsupported format %d\n", buffer->format);
-		return false;
+		return 0;
 	}
-	abort();
 }
 
-static cairo_format_t get_cairo_format(enum wl_shm_format wl_fmt) {
-	switch (wl_fmt) {
-	case WL_SHM_FORMAT_ARGB8888:
-		return CAIRO_FORMAT_ARGB32;
-	case WL_SHM_FORMAT_XRGB8888:
-		return CAIRO_FORMAT_RGB24;
-	default:
-		return CAIRO_FORMAT_INVALID;
+static void compute_composite_region(const struct pixman_f_transform *out2com,
+		int output_width, int output_height, struct grim_box *dest,
+		bool* grid_aligned) {
+	struct pixman_transform o2c_fixedpt;
+	pixman_transform_from_pixman_f_transform(&o2c_fixedpt, out2com);
+
+	struct pixman_vector corners[4];
+	memset(corners, 0, sizeof(corners));
```

Comment:
Maybe use `corners[4] = {0}` instead? Or even `corners[4] = { /* initializer list */ }`?

Reply:
Switched to an initializer list.